atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampUtils.java

```diff
@@ -0,0 +1,244 @@
+
+    private CassandraTimestampUtils() {
+        // utility class
+    }
+
+    public static ByteBuffer constructCheckAndSetMultipleQuery(Map<String, Pair<byte[], byte[]>> checkAndSetRequest) {
+        StringBuilder builder = new StringBuilder();
+        builder.append("BEGIN UNLOGGED BATCH\n"); // Safe, because all updates are on the same partition key
+
+        // Safe, because ordering does not apply in batches
+        checkAndSetRequest.entrySet().forEach(entry -> {
+            String columnName = entry.getKey();
+            byte[] expected = entry.getValue().getLhSide();
+            byte[] target = entry.getValue().getRhSide();
+            builder.append(constructCheckAndSetQuery(columnName, expected, target));
+        });
+        builder.append("APPLY BATCH;");
+        return toByteBuffer(builder.toString());
+    }
+
+    public static boolean isValidTimestampData(byte[] data) {
+        return data != null && data.length == Long.BYTES;
+    }
+
+    public static ByteBuffer constructSelectFromTimestampTableQuery() {
+        return toByteBuffer(String.format(
+                "SELECT %s, %s FROM %s WHERE key=%s;",
+                COLUMN_NAME_COLUMN,
+                VALUE_COLUMN,
+                wrapInQuotes(AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName()),
+                ROW_AND_COLUMN_NAME_HEX_STRING));
+    }
+
+    public static Map<String, byte[]> getValuesFromSelectionResult(CqlResult result) {
+        return result.getRows().stream()
+                .map(CqlRow::getColumns)
+                .collect(Collectors.toMap(
+                        cols -> PtBytes.toString(getNamedColumnValue(cols, COLUMN_NAME_COLUMN)),
+                        cols -> getNamedColumnValue(cols, VALUE_COLUMN)));
+    }
+
+    /**
+     * This method checks that a result from a Cassandra lightweight transaction is either
+     *
+     *  (a) a successful application of proposed changes, or
+     *  (b) an unsuccessful application, but the state of the world in Cassandra as indicated by the casResult
+     *      matches the casMap.
+     *
+     * Otherwise, it throws an IllegalStateException.
+     *
+     * @param casResult the CAS result to verify
+     * @param casMap the intended state of the world
+     * @throws IllegalStateException if the transaction was not successful, and the state of the world does not match
+     *         the intended state, as indicated in the casMap
+     */
+    public static void verifyCompatible(CqlResult casResult, Map<String, Pair<byte[], byte[]>> casMap) {
+        if (!isSuccessfullyApplied(casResult)) {
+            Set<Incongruency> incongruencies = getIncongruencies(casResult, casMap);
+            if (!incongruencies.isEmpty()) {
+                throw new IllegalStateException("[BACKUP/RESTORE] Observed incongruent state of the world; the"
+                        + " following incongruencies were observed: " + incongruencies);
+            }
```

Comment:
That makes sense. Might want a comment of the form "if there are no incongurencies, then the state of the world is what we wanted it to be - possibly because we were upgrading and this call was made in parallel. Since the DB is in the desired state, we're continuing here", and a log.debug message that says "CAS failed, but the DB state is as desired - continuing".